### PR TITLE
LFU Cache Fixes

### PR DESCRIPTION
**Pull request by vmg**
## Description

Some bug fixes, refactorings and more benchmarks for the LFU implementation. Getting it in better shape before we switch to it by default.

## Related Issue(s)
- https://github.com/vitessio/vitess/pull/7439

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by deepthi** [comment]
One thing I missed while reviewing the earlier PR was the names of the flags:
```
  -gate_query_cache_lfu
  -gate_query_cache_memory int
  -gate_query_cache_size int
```
There is no need to prefix the flag names with `gate`. Do you mind renaming them?

**Comment by vmg** [comment]
@deepthi: `-gate_query_cache_size int` is not a new flag, it exists in the current `vtgate`, which is the reason why the two new flags I introduced are also prefixed with `gate_`. Changing `gate_query_cache_size` would be a breaking change. Should I change it anyway? Or change only the two new flags?

File: go/vt/vttablet/tabletserver/query_engine.go [hunk 1]
```diff
@@ -413,6 +413,12 @@ func (qe *QueryEngine) QueryPlanCacheCap() int {
 	return int(qe.plans.MaxCapacity())
 }
 
+// QueryPlanCacheLen returns the length (size in entries0 of the query cache
```
Comment (systay) [inline go/vt/vttablet/tabletserver/query_engine.go:416]: nit: typo. `entries0`

**Comment by vmg** [comment]
OK, fixed and fixed all the flappy tests. I think this PR, which enables the LFU cache by default, is ready to merge. :sparkles:

**Comment by deepthi** [comment]
> @deepthi: `-gate_query_cache_size int` is not a new flag, it exists in the current `vtgate`, which is the reason why the two new flags I introduced are also prefixed with `gate_`. Changing `gate_query_cache_size` would be a breaking change. Should I change it anyway? Or change only the two new flags?

Oops, I did not realize that. I would not hold up this PR. We can rename the flag in a backwards-compatible way in a separate PR (if we decide it is worth doing).

File: go/vt/vttablet/endtoend/config_test.go [hunk 2]
```diff
@@ -203,22 +203,23 @@ func testQueryPlanCache(t *testing.T, cachedPlanSize, cachePlanSize2 int) {
 	client := framework.NewClient()
 	_, _ = client.Execute("select * from vitess_test where intval=:ival1", bindVars)
 	_, _ = client.Execute("select * from vitess_test where intval=:ival2", bindVars)
-	time.Sleep(100 * time.Millisecond)
+	assert.Equal(t, 1, framework.Server.QueryPlanCacheLen())
 
 	vend := framework.DebugVars()
-	verifyIntValue(t, vend, "QueryCacheLength", 1)
+	// verifyIntValue(t, vend, "QueryCacheLength", 1)
```
Comment (systay) [inline go/vt/vttablet/endtoend/config_test.go:209]: did you leave this commented intentionally?

**Review by systay — APPROVED** [review]

File: go/cache/cache.go [hunk 3]
```diff
@@ -76,5 +76,5 @@ type Config struct {
 var DefaultConfig = &Config{
 	MaxEntries:     5000,
 	MaxMemoryUsage: 32 * 1024 * 1024,
-	LFU:            false,
+	LFU:            true,
```
Comment (systay) [inline go/cache/cache.go:79]: 👍

File: go/cache/ristretto/bloom/bbloom.go [hunk 4]
```diff
@@ -43,26 +40,16 @@ func getSize(ui64 uint64) (size uint64, exponent uint64) {
 	return size, exponent
 }
 
-func calcSizeByWrongPositives(numEntries, wrongs float64) (uint64, uint64) {
-	size := -1 * numEntries * math.Log(wrongs) / math.Pow(float64(0.69314718056), 2)
-	locs := math.Ceil(float64(0.69314718056) * size / numEntries)
-	return uint64(size), uint64(locs)
+// NewBloomFilterWithErrorRate returns a new bloomfilter with optimal size for the given
+// error rate
+func NewBloomFilterWithErrorRate(numEntries uint64, wrongs float64) *Bloom {
```
Comment (systay) [inline go/cache/ristretto/bloom/bbloom.go:45]: 👍

File: go/vt/vttablet/endtoend/config_test.go (hunk 2 above)
Comment (vmg) [reply go/vt/vttablet/endtoend/config_test.go:209]: Nope my bad

**Pull request merged**